subs/pantry/src/Pantry/Archive.hs

```diff
@@ -387,41 +390,54 @@ parseArchive pli archive fp = do
             case Map.lookup (seSource se) blobs of
               Nothing -> error $ "Impossible: blob not found for: " ++ seSource se
               Just blobKey -> pure (sfp, TreeEntry blobKey (seType se))
-
           -- parse the cabal file and ensure it has the right name
-          (cabalPath, cabalEntry@(TreeEntry cabalBlobKey _)) <- findCabalFile pli tree
-          mbs <- withStorage $ loadBlob cabalBlobKey
+          (buildFilePath, buildFileEntry@(TreeEntry buildFileBlobKey _), buildFileType) <- findCabalOrHpackFile pli tree
+          mbs <- withStorage $ loadBlob buildFileBlobKey
           bs <-
             case mbs of
-              Nothing -> throwIO $ TreeReferencesMissingBlob pli cabalPath cabalBlobKey
+              Nothing -> throwIO $ TreeReferencesMissingBlob pli buildFilePath buildFileBlobKey
               Just bs -> pure bs
-          (_warnings, gpd) <- rawParseGPD (Left pli) bs
+          cabalBs <- if buildFileType == CabalFile
+                     then return bs
+                     else snd <$> hpackToCabal bs
+          (_warnings, gpd) <- rawParseGPD (Left pli) cabalBs
           let ident@(PackageIdentifier name _) = package $ packageDescription gpd
-          when (cabalPath /= cabalFileName name) $
-            throwIO $ WrongCabalFileName pli cabalPath name
+          when (buildFilePath /= cabalFileName name && buildFileType == CabalFile) $
+            throwIO $ WrongCabalFileName pli buildFilePath name
 
           -- It's good! Store the tree, let's bounce
-          (_tid, treeKey) <- withStorage $ storeTree ident tree cabalEntry
+          (_tid, treeKey) <- withStorage $ storeTree ident tree buildFileEntry buildFileType
+          let packageCabal = if buildFileType == CabalFile
+                             then PCCabalFile buildFileEntry
+                             else PCHpack buildFileEntry
           pure Package
             { packageTreeKey = treeKey
             , packageTree = tree
-            , packageCabalEntry = cabalEntry
+            , packageCabalEntry = packageCabal
             , packageIdent = ident
             }
 
-findCabalFile
+findCabalOrHpackFile
   :: MonadThrow m
   => PackageLocationImmutable -- ^ for exceptions
   -> Tree
-  -> m (SafeFilePath, TreeEntry)
-findCabalFile loc (TreeMap m) = do
+  -> m (SafeFilePath, TreeEntry, BuildFileType)
+findCabalOrHpackFile loc (TreeMap m) = do
   let isCabalFile (sfp, _) =
         let txt = unSafeFilePath sfp
-         in not ("/" `T.isInfixOf` txt) && ".cabal" `T.isSuffixOf` txt
-  case filter isCabalFile $ Map.toList m of
+         in not ("/" `T.isInfixOf` txt) && (".cabal" `T.isSuffixOf` txt)
+      isHpackFile (sfp, _) =
+        let txt = unSafeFilePath sfp
+         in T.pack (Hpack.packageConfig) == txt
+  case filter (\f -> isCabalFile f || isHpackFile f) $ Map.toList m of
```

Comment:
How about `mapMaybe` and have a function that categorizes the files as cabal or hpack files, so that we don't have to call `isHpackFile` again? Also, instead of having a separate `BuildFileType`, I think I'd prefer to see:

```haskell
data BuildFile
  = BFCabal !SafeFilePath !TreeEntry
  | BFHpack !TreeEntry
```

Note that we don't even need a `SafeFilePath` for hpack files since we know what it has to be (package.yaml).

Reply:
Done.